### ISSUE TITLE
App Directory: Unable to Redirect from middleware when having basePath and locales

### ISSUE THREAD

**Issue by alikleit** [vercel/next.js#50161]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 10 Pro
    Binaries:
      Node: 18.15.0
      npm: N/A
      Yarn: N/A
      pnpm: 7.29.1
    Relevant packages:
      next: 13.4.4-canary.0
      eslint-config-next: 13.4.3
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.0.4
```

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/alikleit/next-middleware-issue

### To Reproduce

Following the documentation from next internationalization using `App Directory`, with adding a middleware to redirect from the root `/` to a specific locale `/en` . This work perfectly until adding a `basePath` :
```
// next.config.js
const nextConfig = {
  basePath: "/admin",
};
```

Starting the app in dev mode `pnpm dev` and redirect to `/admin`, middleware wont be triggered.

### Describe the Bug

Middleware is not triggered when in the root after setting basePath, with a Next app that uses only App Directory, and localized.

### Expected Behavior

Middleware should trigger on the root of the application using basePath. and redirect to a specifc locale

### Which browser are you using? (if relevant)

Edge

**Comment by petmarsh** [comment]
I had a similar problem with middleware - see my reported [issue](https://github.com/vercel/next.js/issues/47085)

Luckily there's a workaround (did not find yet a problem with that until now while developing/testing), give it a try just by adding 
{ source: '/' },
into the matcher list, hopefully works on your platform as well

**Comment by petmarsh** [comment]
> Works! had to change the response redirect url to add the `basePath` as well, though is this property mentioned somewhere in their docs?

yeah, there are a few tweaks required, but at least the / path is matched
and regarding docs not really, that's the reason for me logging a doc update issue - in my case it was more in the direction of merging info from various places and getting into the nextjs source, something like "help yourself, you won't get any help" ;)

**Comment by kjxbyz** [comment]
it doesn't work for me.

`next.config.js`

```
/** @type {import('next').NextConfig} */
const nextConfig = {
  basePath: "/test",
  trailingSlash: true,
  reactStrictMode: true,
  swcMinify: true,
};

module.exports = nextConfig;
```

`middleware.ts`

```
export const config = {
  // Matcher ignoring `/_next/` and `/api/`
  matcher: [
    '/((?!api|_next/static|_next/image|favicon.ico).*)',
    { source: '/' },
  ],
}
```

**Comment by kjxbyz** [comment]
good work.

**Comment by matteopaoli** [comment]
This isn't scalable though, what if I want to apply the same logic on every path instead of just `/`? 

I've run an attempt by using a template string and inserting the basePath at the beginning of the matcher string, but it's not working as expected

**Comment by matteopaoli** [comment]
Seems like the best way to handle this case is replacing the `config` object with an if statement inside the middleware. Leaving the code with an helper function in case someone has to migrate the approach too.

`function doesNotStartWithAny(value: string, substrings: string[]): boolean {
  return !substrings.some((substring) => value.startsWith(substring))
}

export async function middleware(req: NextRequest): Promise<NextResponse | undefined> {
    if (
    doesNotStartWithAny(req.nextUrl.pathname, [
      '/api',
      '/_next/static',
      '/_next/image',
      '/assets',
      '/favicon.ico',
      '/images',
      '/manifest',
    ])
  ) {
   // middleware logic...
  }
  return NextResponse.next()
}
`

**Comment by reghuramk** [comment]
Hi, 
I'm unable to continue my routing in the middleware which contains a locale prefix in the URL.
if I navigate to localhost:/homepage which is not inside the [locale] folder, the middleware is able to successfully continue the routing.

But if I navigate to a route which is inside the [locale] folder localhost:/en/login, the middleware redirects me to a not found page.
I have disabled the localeprefix in the middleware config.

My middleware configuration

`import createMiddleware from "next-intl/middleware";
import { NextRequest, NextResponse } from "next/server";

export default createMiddleware({
  // A list of all locales that are supported
  locales: ["en", "de"],

  // Used when no locale matches
  defaultLocale: "en",
  localePrefix: "never",
  localeDetection: true,
});

export const config = {
  // Match only internationalized pathnames
  matcher: [
    // Match all pathnames except for
    // - … if they start with `/api`, `/_next` or `/_vercel`
    // - … the ones containing a dot (e.g. `favicon.ico`)
    "/((?!api|_next|_vercel|.*\\..*).*)",
    // However, match all pathnames within `/users`, optionally with a locale prefix
    "/([\\w-]+)?/users/(.+)",
  ],
};

export function middleware(request: NextRequest) {
  return NextResponse.next();
}
`

**Comment by reghuramk** [comment]
Hi,
I'm unable to continue my routing in the middleware which contains a locale prefix in the URL.
if I navigate to localhost:/homepage which is not inside the [locale] folder, the middleware is able to successfully continue the routing.

But if I navigate to a route which is inside the [locale] folder localhost:/en/login, the middleware redirects me to a not found page.
I have disabled the localeprefix in the middleware config.

My middleware configuration

```
export default createMiddleware({
 locales: ["en", "de"],defaultLocale: "en",
  localePrefix: "never",
   localeDetection: true,
 )}
 
 export const config = {
  // Match only internationalized pathnames
  matcher: [
    "/((?!api|_next|_vercel|.*\\..*).*)",
    "/([\\w-]+)?/users/(.+)",
  ],
};

export function middleware(request: NextRequest) {
  return NextResponse.next();
}

```

**Comment by arnaud-deprez** [comment]
Same issue with app router and NextJS `14.2.4`.

To add some extra information

The middleware on root page `/${basePath}` is not triggered as soon as we add a matcher with regular expression to exclude some path.

With eg:
```ts
// next.config.mjs
/** @type {import('next').NextConfig} */
const nextConfig = {
  basePath: '/admin',
};

// middleware.ts
export const config = { matcher: ["/((?!api/foo).*)"] };
```
Then, the middleware is not triggered when we navigate to `/admin`

However, if we add `/` in the middleware configuration:
```ts
export const config = { matcher: ["/((?!api/foo).*)", "/"] };
```
Then it works.

This is unexpected because the regex should match `/` too.

**Comment by ChenAlmost** [comment]
I have the same issue

**Comment by marcoshernanz** [comment]
Thanks for the report. This still reproduces on the latest canary, but it’s the same bare-basePath matcher bug tracked in #47085 (see also #73786), so I’m closing this as a duplicate. Adding `/` explicitly to the matcher remains the workaround.
